Adds parallel execution of actions in stages

JayjeetAtGithub opened the pull request:
Runs the actions in a stage in parallel

JayjeetAtGithub commented:
@ivotron Can you please help me on how to add the `futures` dependency to the codebase. I added it to `setup.py` first , but build failed. Then i tried sometime more, but still failing.

JayjeetAtGithub commented:
```
$ bash -c "popper run --wfile ci/cli.workflow"
Error: Command 'run' doesn't exist. 
Type 'popper --help' for more.
The most similar commands are: run, version
No module named concurrent.futures
The command "bash -c "popper run --wfile ci/cli.workflow"" exited with 2.
```

This error comes during python2.7 build while python3.7 passes

ivotron commented:
sorry @JayjeetAtGithub can you please rebase master?

JayjeetAtGithub commented:
@ivotron  I have rebased, but my python2.7 build is failing ? Can you please help a bit

ivotron commented on .gitmodules:
```diff
@@ -31,3 +31,6 @@
 	url = https://github.com/PythonCharmers/python-future.git
+[submodule "cli/bin/vendor/pythonfutures"]
+	path = cli/bin/vendor/pythonfutures
+	url = https://github.com/agronholm/pythonfutures.git
```
the `futures` module is already being added, so there's no need to add it again. This was done as part of #524

ivotron commented on cli/bin/popper:
```diff
@@ -16,6 +16,7 @@ else:
 sys.path.insert(0, yaml_lib)
 if sys.version_info[0] < 3:
     sys.path.insert(0, os.path.join(vendordir, 'python-future', 'src'))
+    sys.path.insert(0, os.path.join(vendordir, 'pythonfutures', 'src'))
```
same as above

ivotron replied:
@JayjeetAtGithub this does not work because the `pythonfeatures` repo does not have a `src` folder. The library is at the root of the project

ivotron commented on cli/setup.py:
```diff
@@ -20,7 +20,8 @@
         'pyyaml',
-        'pyaes'
+        'pyaes',
+        'futures; python_version == "2.7"'
```
same as above

JayjeetAtGithub commented:
@ivotron The module that i added as a part of this pr is [futures](https://pypi.org/project/futures/) which is different from that added in #524.I removed it, but build is still failing.

ivotron commented:
@JayjeetAtGithub oh, I see. In that case, what you had before should work. I left a comment in one of our threads.

JayjeetAtGithub commented:
@ivotron Thanks a lot for pointing it out. Updated the PR

ivotron commented:
I hate to ask this again: could you please rebase master again? Sorry 😩

JayjeetAtGithub commented:
@ivotron Added --serial flag

JayjeetAtGithub commented:
Thanks ! Will be updating soon

JayjeetAtGithub commented:
@ivotron Added tests to check thread safety with stages each containing 10 actions.

ivotron commented:
@JayjeetAtGithub thanks a lot for addressing the previous comments!

The output on TravisCI seems to indicate that only the test for `reuse` is being executed, can you please check? One thing I suspect it will be an issue is the printing of messages to the terminal. We might need to serialize access to stdout within the `utils.exec_cmd` function. In general, we need to check all the codebase to ensure that there are no thread-safety issues. Another one from the top of my head is the `Workflow.get_stages` generator method, which will likely require the use of thread-safe iterators.

Also, we will need to run the tests on travis using the `--serial` flag (i.e. we need to modify the `.travis.yml` so that this flag is passed), otherwise all the tests we run will be implicitly testing this functionality as well. Another alternative is to invert the behavior and add a `--parallel` flag and always run serially by default.

JayjeetAtGithub commented:
> One thing I suspect it will be an issue is the printing of messages to the terminal. We might need to serialize access to stdout within the utils.exec_cmd function. 

@ivotron Can you explain this a bit ?

JayjeetAtGithub commented:
> The output on TravisCI seems to indicate that only the test for reuse is being executed, can you please check?

Yes , only reuse is getting executed . Sorry for not noticing it before.

pep8speaks commented:
Hello @JayjeetAtGithub! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`cli/popper/commands/cmd_run.py`](https://github.com/systemslab/popper/blob/151c0dd0412fe3df52ceac39f2255fb6e4e43a4b/cli/popper/commands/cmd_run.py):

> [Line 64:80](https://github.com/systemslab/popper/blob/151c0dd0412fe3df52ceac39f2255fb6e4e43a4b/cli/popper/commands/cmd_run.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 75:80](https://github.com/systemslab/popper/blob/151c0dd0412fe3df52ceac39f2255fb6e4e43a4b/cli/popper/commands/cmd_run.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

ivotron commented:
> > One thing I suspect it will be an issue is the printing of messages to the terminal. We might need to serialize access to stdout within the utils.exec_cmd function.
> 
> @ivotron Can you explain this a bit ?

sure! Currently, [this function](https://github.com/systemslab/popper/blob/1f04249de147b7f46517531f070f7ea33827dec3/cli/popper/utils.py#L89) makes reference to `stdout` via the `info()` function. When `popper run` is executed (with `verbose=True` and `log_file=True`), multiple threads executing this function will be printing lines of output that can correspond to multiple threads. All the output will likely end up being interleaved.

One pragmatic solution is to notify users that this is likely to happen when they execute in parallel mode, and that we advice them to use `--quiet` to avoid confusion.

JayjeetAtGithub commented:
@ivotron  Summary of changes made:

- Changed `--serial` to `--parallel` and changed the tests accordingly .
- Made the generator function `Workflow.get_stages` thread safe by making a custom decorator `@threadsafe_generator` which is defined in `utils.py` file.
- The above change resulted in execution of all tests  and proper output in build logs which was not happening before.
- Added a warning message about interleaved output on `popper run --parallel`.

JayjeetAtGithub commented:
@ivotron I cant understand why some of these actions are failing silently. Can you please tell me ?

JayjeetAtGithub commented:
Hi , @ivotron I have modified the `ThreadPoolExecutor` part in a way, so that if any one of the actions in a stage ( where all the actions are running parallely ) do fail, then the execution stops with `exit code 1` and hence the build fails. I have also corrected the tests, previously due to which all the `[ls...]` actions failed.

JayjeetAtGithub commented:
@ivotron Ok !  Updating soon !

JayjeetAtGithub commented:
@ivotron Made the changes accordingly. Added tests to check whether all the required `.log` files are being generated in `popper_logs` directory in the workspace for every action.

JayjeetAtGithub commented:
Hi @ivotron, Changes made :
- Fixed PEP8 issues
- Modified the `run_stage()` fn. by adding `pu.info('Action ran successfully !')` whenever an action runs successfully in parallel mode. Then i added tests to match the count of this info statement in output with the total no. of actions. Hope this serves the purpose. :smiley:

JayjeetAtGithub commented:
@ivotron Added failure tests

JayjeetAtGithub commented on cli/popper/gha.py:
```diff
@@ -290,12 +292,30 @@ def run(self, action_name=None, reuse=False):
+            with ThreadPoolExecutor(max_workers=mp.cpu_count()) as ex:
+                flist = {
+                    ex.submit(
+                        self.stage_exec,
```
Done

JayjeetAtGithub commented on cli/popper/gha.py:
```diff
@@ -290,12 +292,30 @@ def run(self, action_name=None, reuse=False):
+                    ex.submit(
+                        self.stage_exec,
+                        a,
+                        reuse): a for a in stage}
```
Done

JayjeetAtGithub commented on cli/setup.py:
```diff
@@ -20,7 +20,8 @@
         'pyyaml',
-        'pyaes'
+        'pyaes',
+        'futures; python_version == "2.7"'
```
Done

ivotron commented:
thanks a lot @JayjeetAtGithub !
